aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,606 @@
+                // Aggregator knows the value, therefore we can subtract
+                // checking we don't drop below zero.
+                let result = subtraction(self.value, value)?;
+                self.value = result;
+                Ok(())
+            }
+            // Since `sub` is a barrier, we never encounter delta during
+            // subtraction. This will change in future.
+            // TODO: implement this properly once `sub` stops being a barrier.
+            AggregatorState::PositiveDelta => {
+                unreachable!("subtraction always materializes the value")
+            }
+        }
+    }
+
+    /// Materializes the value of the aggregator, changing its state.
+    fn materialize(
+        &mut self,
+        context: &NativeAggregatorContext,
+        handle: &TableHandle,
+    ) -> PartialVMResult<()> {
+        // If aggregator has already been materialized, return immediately.
+        if self.state == AggregatorState::Data {
+            return Ok(());
+        }
+
+        // Otherwise, we have a delta and have to go to storage.
+        let key_bytes = serialize(&self.key);
+        match context.resolver.resolve_table_entry(handle, &key_bytes) {
+            Err(_) => Err(extension_error("error resolving aggregator's value")),
+            Ok(maybe_bytes) => {
+                match maybe_bytes {
+                    Some(bytes) => {
+                        let base = deserialize(&bytes);
+
+                        // The only remaining case is PositiveDelta. But assert just in case.
+                        assert!(self.state == AggregatorState::PositiveDelta);
+
+                        // At this point, we fetched the "true" value of the
+                        // aggregator from the previously executed transaction.
+                        let result = addition(base, self.value, self.limit)?;
+                        self.value = result;
+                        self.state = AggregatorState::Data;
+                        Ok(())
+                    }
+                    None => Err(extension_error("aggregator's value not found")),
+                }
+            }
+        }
+    }
+}
+
+/// Aggregator table - a top-level collection storing aggregators. It holds an
+/// inner table data structure to actually store the aggregators.
+struct AggregatorTable {
+    // All aggregators created with this table.
+    new_aggregators: BTreeSet<u128>,
+    // All aggregators removed from this table.
+    removed_aggregators: BTreeSet<u128>,
+    // Table to store all agregators associated with this `AggregatorTable`.
+    table: BTreeMap<u128, Aggregator>,
+}
+
+impl AggregatorTable {
+    /// Returns a possibly new aggregator instance. If transaction did not
+    /// initialize this aggregator, then the actual value is not known, and the
+    /// aggregator state must be a delta, zero-intialized.
+    fn get_or_create_aggregator(&mut self, key: u128, limit: u128) -> &mut Aggregator {
```

Comment:
If I am right, this never creates an actually new aggregator ("create_new_aggregator" does that), but comment "returns a possibly new aggregator instance" and "create" in the name may give a wrong impression.

maybe calling it get_existing_aggregator can be more descriptive, or anything that avoids this potential confusion.

Reply:
I think what causes confusion is that there is Move aggregator - a single entity, and a rust `Aggregator` - struct that represents the state of the Move aggregator in a a given context/transaction. We do create this struct in Rust (but yes, not the Move one)

Reply:
Makes sense.